picoquic/frames.c

```diff
@@ -24,6 +24,129 @@
 #include <stdlib.h>
 #include <string.h>
 
+/* ****************************************************
+ * Frames private declarations
+ * ****************************************************/
+
+static const size_t challenge_length = 8;
+
+
+/**
+ * Frame decoder function
+ * Inputs:
+ *   cnx       - [in/out] picoquic Context
+ *   bytes     - [in]     pointer to the beginning of the frame (frame type)
+ *   bytes_max - [in]     pointer to the end of the packet (one past the last byte)
+ * Returns:
+ *   Pointer to the data following the end of this frame, if the frame has been decoded successfully;
+ *   or NULL if, decoding failed (in which case, picoquic_connection_error has been called).
+ */
+typedef uint8_t* (*decode_frame_fn)(picoquic_cnx_t* cnx, uint8_t* bytes, const uint8_t* bytes_max);
+
+/**
+ * Frame skip function
+ * Inputs:
+ *   bytes     - [in]     pointer to the beginning of the frame (frame type)
+ *   bytes_max - [in]     pointer to the end of the packet (one past the last byte)
+ * Returns:
+ *   Pointer to the data following the end of this frame, if the frame has been skipped successfully;
+ *   or NULL if, skipping failed.
+ */
+typedef uint8_t* (*skip_frame_fn)(uint8_t* bytes, const uint8_t* bytes_max);
+
+
+/* ****************************************************
+ * Helper utilities
+ * ****************************************************/
+
+/* Skip and decode function.
+ * These functions return NULL in case of a failure (insufficient buffer).
+ */
+
+#define VARINT_LEN(bytes) (1U << (((bytes)[0] & 0xC0) >> 6))
+
+
+static uint8_t* picoquic_frames_fixed_skip(uint8_t* bytes, const uint8_t* bytes_max, size_t size)
+{
+    return (bytes += size) <= bytes_max ? bytes : NULL;
+}
+
+
+static uint8_t* picoquic_frames_varint_skip(uint8_t* bytes, const uint8_t* bytes_max)
+{
+    return bytes < bytes_max ? picoquic_frames_fixed_skip(bytes, bytes_max, VARINT_LEN(bytes)) : NULL;
+}
+
+
+/* Parse a varint. In case of an error, *n64 is unchanged, and NULL is returned */
+static uint8_t* picoquic_frames_varint_decode(uint8_t* bytes, const uint8_t* bytes_max, uint64_t* n64)
+{
+    uint8_t length;
+
+    if (bytes < bytes_max && bytes + (length=VARINT_LEN(bytes)) <= bytes_max) {
+        uint64_t v = *bytes++ & 0x3F;
+
+        while (--length > 0) {
+            v <<= 8;
+            v += *bytes++;
+        }
+
+        *n64 = v;
+    } else {
+        bytes = NULL;
+    }
+
+    return bytes;
+}
+
+
+static uint8_t* picoquic_frames_uint8_decode(uint8_t* bytes, const uint8_t* bytes_max, uint8_t* n)
+{
+    if (bytes < bytes_max) {
+        *n = *bytes++;
+    } else {
+        bytes = NULL;
+    }
+    return bytes;
+}
+
+
+static uint8_t* picoquic_frames_uint16_decode(uint8_t* bytes, const uint8_t* bytes_max, uint16_t* n)
+{
+    if (bytes + sizeof(*n) <= bytes_max) {
+        *n = PICOPARSE_16(bytes);
+        bytes += sizeof(*n);
+    } else {
+        bytes = NULL;
+    }
+    return bytes;
+}
+
```

Comment:
Same remark as varint -- less important as this is trivial, but having an unit tests would be nice.